src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1023,195 +1018,224 @@ static ucs_status_t uct_ib_mlx5_devx_umr_modify_qp(uct_ib_mlx5_md_t *md)
-    if (NULL == umr->umr_mkey) {
-        ucs_error("%s: failed to create UMR mkey: %m",
-                  uct_ib_device_name(&md->super.dev));
-        ucs_free(umr);
-        return NULL;
-    }
-
-    ucs_trace("%s: created " UCT_IB_MLX5_UMR_FMT,
-              uct_ib_device_name(&md->super.dev), UCT_IB_MLX5_UMR_ARG(umr));
-
-    return umr;
-}
-
-static void uct_ib_mlx5_devx_umr_mkey_destroy(uct_ib_mlx5_md_t *md,
-                                              uct_ib_mlx5_devx_umr_t *umr)
+static ucs_status_t uct_ib_mlx5_devx_umr_init_import(uct_ib_mlx5_md_t *md)
 {
-    ucs_trace("%s: destroy " UCT_IB_MLX5_UMR_FMT,
-              uct_ib_device_name(&md->super.dev), UCT_IB_MLX5_UMR_ARG(umr));
-
-    if (mlx5dv_destroy_mkey(umr->umr_mkey) != 0) {
-        ucs_error("%s: failed to destroy " UCT_IB_MLX5_UMR_FMT ": %m",
-                  uct_ib_device_name(&md->super.dev), UCT_IB_MLX5_UMR_ARG(umr));
+    /* Initialize indirect UMR mkey hash table */
+    md->umr.mkey_hash = kh_init(umr_mkey_map);
+    if (md->umr.mkey_hash == NULL) {
+        return UCS_ERR_NO_MEMORY;
     }
 
-    ucs_free(umr);
-}
-
-static void
-uct_ib_mlx5_devx_umr_mkey_alias_destroy(uct_ib_mlx5_md_t *md,
-                                        uct_ib_mlx5_devx_umr_alias_t *umr_alias)
-{
-    ucs_status_t status;
-
-    ucs_trace("%s: destroy " UCT_IB_MLX5_UMR_ALIAS_FMT,
-              uct_ib_device_name(&md->super.dev),
-              UCT_IB_MLX5_UMR_ALIAS_ARG(umr_alias));
-
-    status = uct_ib_mlx5_devx_obj_destroy(umr_alias->cross_mr, "MKEY_ALIAS");
-    if (UCS_OK != status) {
-        ucs_error("%s: failed to destroy " UCT_IB_MLX5_UMR_ALIAS_FMT ": %m",
-                  uct_ib_device_name(&md->super.dev),
-                  UCT_IB_MLX5_UMR_ALIAS_ARG(umr_alias));
-    }
+    return UCS_OK;
 }
 
-static void uct_ib_mlx5_devx_umr_free(uct_ib_mlx5_md_t *md)
+static void uct_ib_mlx5_devx_umr_cleanup(uct_ib_mlx5_md_t *md)
 {
-    uct_ib_device_t *ibdev = &md->super.dev;
     uct_ib_mlx5_devx_umr_alias_t mkey_alias;
-    uct_ib_mlx5_devx_umr_t *item, *tmp;
+    uct_ib_mlx5_devx_umr_mkey_t *item, *tmp;
+    int ret;
 
     /* Destroy UMR mkey hash if present */
-    if (NULL != md->umr_mkey_hash) {
+    if (NULL != md->umr.mkey_hash) {
         ucs_trace("%s: destroy UMR mkey hash with %d elements",
-                  uct_ib_device_name(ibdev), kh_size(md->umr_mkey_hash));
+                  uct_ib_mlx5_dev_name(md), kh_size(md->umr.mkey_hash));
 
-        kh_foreach_value(md->umr_mkey_hash, mkey_alias, {
+        kh_foreach_value(md->umr.mkey_hash, mkey_alias, {
             uct_ib_mlx5_devx_umr_mkey_alias_destroy(md, &mkey_alias);
         });
 
-        kh_destroy(umr_mkey_map, md->umr_mkey_hash);
-        md->umr_mkey_hash = NULL;
+        kh_destroy(umr_mkey_map, md->umr.mkey_hash);
     }
 
     /* Destroy UMR mkey pool if not empty */
-    if (!ucs_list_is_empty(&md->umr_mkey_pool)) {
+    if (!ucs_list_is_empty(&md->umr.mkey_pool)) {
```

Comment:
we can remove is_empty check and the debug trace, since having non-empty pool is not a leak